File: tools/xtask/src/interop.rs
//! Cross-language interop orchestrator (`cargo run -p xtask interop`).
//!
//! For each requested language it generates the interop package, builds that
//! language's harness, then runs the full matrix: every client language against
//! every server language over a per-server loopback port (TCP for rpc/events, UDP
//! for datagrams), for each transport, and reports per-cell pass/fail. See
//! tests/interop/README.md.

use std::io::{BufRead, BufReader, Read};
use std::path::{Path, PathBuf};
use std::process::{Child, Command, Stdio};
use std::time::{Duration, Instant};

use anyhow::{Context, Result, bail};

#[derive(Clone, Copy, PartialEq)]
enum BuildKind {
    Cargo,
    Go,
    None,
    /// A self-describing harness: `build.sh` builds it and `run <mode> <transport>
    /// <socket>` runs it (the script owns its own toolchain env). Used by every
    /// language past the rust/go/python reference trio so wiring is uniform.
    Script,
}

#[derive(Clone)]
struct Lang {
    name: &'static str,
    gen_target: &'static str,
    package_name: &'static str,
    harness_dir: &'static str,
    gen_subdir: &'static str,
    build: BuildKind,
    /// The loopback port this language's SERVER binds (clients connect here). One per
    /// language; reused across the three transports, which run sequentially.
    port: u16,
}

/// All languages with a harness. Add an entry as each language's harness lands.
fn registry() -> Vec<Lang> {
    vec![
        Lang {
            name: "rust",
            gen_target: "rust",
            package_name: "interop_api",
            harness_dir: "tests/interop/harness/rust",
            gen_subdir: "gen",
            build: BuildKind::Cargo,
            port: 6387,
        },
        Lang {
            name: "go",
            gen_target: "go",
            package_name: "interopapi",
            harness_dir: "tests/interop/harness/go",
            gen_subdir: "gen",
            build: BuildKind::Go,
            port: 6388,
        },
        Lang {
            name: "python",
            gen_target: "python",
            package_name: "interop_api",
            harness_dir: "tests/interop/harness/python",
            gen_subdir: "gen",
            build: BuildKind::None,
            port: 6389,
        },
        script_lang("typescript", 6390),
        script_lang("java", 6391),
        script_lang("csharp", 6392),
        script_lang("c", 6393),
        script_lang("ruby", 6394),
        script_lang("elixir", 6395),
        script_lang("dart", 6396),
        script_lang("ocaml", 6397),
        script_lang("zig", 6398),
        script_lang("kotlin", 6399),
        // 6400 is reserved for swift. It is excluded from the matrix for now (no
        // `swiftc` toolchain in this environment); when one is available, add
        // `script_lang("swift", 6400)` here and create tests/interop/harness/swift/.
    ]
}

/// A language whose harness follows the uniform `build.sh` + `run` convention.
fn script_lang(name: &'static str, port: u16) -> Lang {
    // harness_dir leaks the name via a leaked String so it can stay `&'static str`
    // in Lang; the registry is built once per process.
    let dir: &'static str = Box::leak(format!("tests/interop/harness/{name}").into_boxed_str());
    Lang {
        name,
        gen_target: name,
        package_name: "interop_api",
        harness_dir: dir,
        gen_subdir: "gen",
        build: BuildKind::Script,
        port,
    }
}

impl Lang {
    fn dir(&self, root: &Path) -> PathBuf {
        root.join(self.harness_dir)
    }
    fn gen_dest(&self, root: &Path) -> PathBuf {
        self.dir(root).join(self.gen_subdir)
    }

    /// The argv prefix + env that launches this harness; `server`/`client`,
    /// transport, and socket are appended by the caller.
    fn launch(&self, root: &Path) -> (Vec<String>, Vec<(String, String)>) {
        let dir = self.dir(root);
        match self.name {
            "rust" => (
                vec![
                    dir.join("target/debug/csil-interop")
                        .to_string_lossy()
                        .into_owned(),
                ],
                vec![],
            ),
            "go" => (
                vec![dir.join("csil-interop").to_string_lossy().into_owned()],
                vec![],
            ),
            "python" => {
                let pp = format!(
                    "{}:{}",
                    self.gen_dest(root).display(),
                    root.join("transports/python").display()
                );
                (
                    vec![
                        "python3".to_string(),
                        dir.join("main.py").to_string_lossy().into_owned(),
                    ],
                    vec![("PYTHONPATH".to_string(), pp)],
                )
            }
            _ => (vec![dir.join("run").to_string_lossy().into_owned()], vec![]),
        }
    }
}

fn run_in(dir: &Path, program: &str, args: &[&str]) -> Result<()> {
    let ok = Command::new(program)
        .args(args)
        .current_dir(dir)
        .status()
        .with_context(|| format!("failed to launch {program} in {}", dir.display()))?
        .success();
    if !ok {
        bail!("`{program} {}` failed in {}", args.join(" "), dir.display());
    }
    Ok(())
}

/// Generate the interop package for one language into its harness `gen/` dir.
fn generate(root: &Path, cli: &Path, lang: &Lang) -> Result<()> {
    let spec = root.join("tests/interop/interop.csil");
    let spec_body =
        std::fs::read_to_string(&spec).with_context(|| format!("reading {}", spec.display()))?;
    let opts = format!(
        "options {{ emit_packages: [\"{}\"], package_name: \"{}\", package_version: \"0.1.0\" }}\n\n",
        lang.gen_target, lang.package_name
    );
    let gen_input_dir = root.join("target/interop");
    std::fs::create_dir_all(&gen_input_dir)?;
    let gen_input = gen_input_dir.join(format!("gen-input-{}.csil", lang.name));
    std::fs::write(&gen_input, format!("{opts}{spec_body}"))?;

    let dest = lang.gen_dest(root);
    let _ = std::fs::remove_dir_all(&dest);
    let ok = Command::new(cli)
        .args(["generate", "--input"])
        .arg(&gen_input)
        .args(["--target", lang.gen_target, "--output"])
        .arg(&dest)
        .status()
        .context("running csilgen generate")?
        .success();
    if !ok {
        bail!("csilgen generate failed for {}", lang.name);
    }
    Ok(())
}

fn build_lang(root: &Path, lang: &Lang) -> Result<()> {
    let dir = lang.dir(root);
    println!("  building {} harness…", lang.name);
    match lang.build {
        BuildKind::Cargo => run_in(&dir, "cargo", &["build", "--quiet"]),
        BuildKind::Go => run_in(&dir, "go", &["build", "-o", "csil-interop", "."]),
        BuildKind::None => Ok(()),
        BuildKind::Script => {
            // Make the convention scripts executable, then build via build.sh.
            for f in ["build.sh", "run"] {
                let p = dir.join(f);
                if let Ok(meta) = std::fs::metadata(&p) {
                    let mut perm = meta.permissions();
                    use std::os::unix::fs::PermissionsExt;
                    perm.set_mode(0o755);
                    let _ = std::fs::set_permissions(&p, perm);
                }
            }
            run_in(&dir, "bash", &["build.sh"])
        }
    }
}

/// Spawn a server and wait until it prints `READY` on stdout (or time out).
fn spawn_server(
    argv: &[String],
    env: &[(String, String)],
    transport: &str,
    port: u16,
) -> Result<Child> {
    let mut cmd = Command::new(&argv[0]);
    cmd.args(&argv[1..]);
    cmd.arg("server").arg(transport).arg(port.to_string());
    for (k, v) in env {
        cmd.env(k, v);
    }
    cmd.stdout(Stdio::piped()).stderr(Stdio::null());
    let mut child = cmd.spawn().context("spawning server")?;

    let stdout = child.stdout.take().unwrap();
    let mut reader = BufReader::new(stdout);
    let deadline = Instant::now() + Duration::from_secs(20);
    let mut line = String::new();
    loop {
        line.clear();
        if Instant::now() > deadline {
            let _ = child.kill();
            bail!("server did not become READY within timeout");
        }
        let n = reader.read_line(&mut line)?;
        if n == 0 {
            let _ = child.kill();
            bail!("server exited before READY");
        }
        if line.trim() == "READY" {
            break;
        }
    }
    // Drain remaining stdout in the background so the pipe never blocks the server.
    std::thread::spawn(move || {
        let mut sink = Vec::new();
        let _ = reader.read_to_end(&mut sink);
    });
    Ok(child)
}

/// Run a client and parse its JSON result into (all_ok, failed-case details).
fn run_client(
    argv: &[String],
    env: &[(String, String)],
    transport: &str,
    port: u16,
) -> (bool, Vec<String>) {
    let mut cmd = Command::new(&argv[0]);
    cmd.args(&argv[1..]);
    cmd.arg("client").arg(transport).arg(port.to_string());
    for (k, v) in env {
        cmd.env(k, v);
    }
    let out = match cmd.output() {
        Ok(o) => o,
        Err(e) => return (false, vec![format!("spawn failed: {e}")]),
    };
    let stdout = String::from_utf8_lossy(&out.stdout);
    let json_line = stdout
        .lines()
        .rev()
        .find(|l| l.trim_start().starts_with('{'));
    let line = match json_line {
        Some(l) => l,
        None => {
            let err = String::from_utf8_lossy(&out.stderr);
            return (
                false,
                vec![format!("no JSON result (stderr: {})", err.trim())],
            );
        }
    };
    let v: serde_json::Value = match serde_json::from_str(line) {
        Ok(v) => v,
        Err(e) => return (false, vec![format!("bad JSON: {e}")]),
    };
    let mut failures = Vec::new();
    if let Some(cases) = v.get("cases").and_then(|c| c.as_array()) {
        if cases.is_empty() {
            failures.push("no cases run".to_string());
        }
        for c in cases {
            let ok = c.get("ok").and_then(|b| b.as_bool()).unwrap_or(false);
            if !ok {
                let name = c.get("name").and_then(|n| n.as_str()).unwrap_or("?");
                let detail = c.get("detail").and_then(|d| d.as_str()).unwrap_or("");
                failures.push(format!("{name}: {detail}"));
            }
        }
    } else {
        failures.push("malformed result (no cases array)".to_string());
    }
    (failures.is_empty(), failures)
}

pub fn run(langs_filter: Vec<String>, transports_filter: Vec<String>) -> Result<()> {
    let root = std::env::current_dir()?;
    let cli = root.join("target/release/csilgen");

    println!("== building csilgen CLI (release) ==");
    run_in(
        &root,
        "cargo",
        &["build", "--release", "--quiet", "-p", "csilgen"],
    )?;
    if !cli.exists() {
        bail!("csilgen binary not found at {}", cli.display());
    }

    let all = registry();
    let langs: Vec<Lang> = if langs_filter.is_empty() {
        all
    } else {
        all.into_iter()
            .filter(|l| langs_filter.iter().any(|f| f == l.name))
            .collect()
    };
    if langs.is_empty() {
        bail!("no known languages selected");
    }
    let transports: Vec<String> = if transports_filter.is_empty() {
        vec!["rpc".into(), "events".into(), "datagrams".into()]
    } else {
        transports_filter
    };

    println!(
        "== interop: langs [{}] transports [{}] ==",
        langs.iter().map(|l| l.name).collect::<Vec<_>>().join(", "),
        transports.join(", ")
    );

    // Phase 1: generate + build each harness.
    for lang in &langs {
        println!("== preparing {} ==", lang.name);
        generate(&root, &cli, lang)?;
        build_lang(&root, lang)?;
    }

    // Phase 2: run the matrix. The server binds its own loopback port; every client
    // connects to that port. One port per server-lang, reused across transports
    // (they run sequentially; harnesses set SO_REUSEADDR for clean rebind).
    // results[(transport, client, server)] = (ok, failures)
    let mut results: Vec<(String, String, String, bool, Vec<String>)> = Vec::new();
    for transport in &transports {
        for server in &langs {
            let (sargv, senv) = server.launch(&root);
            let mut child = match spawn_server(&sargv, &senv, transport, server.port) {
                Ok(c) => c,
                Err(e) => {
                    // Mark every client cell against this server as failed.
                    for client in &langs {
                        results.push((
                            transport.clone(),
                            client.name.to_string(),
                            server.name.to_string(),
                            false,
                            vec![format!("server start: {e}")],
                        ));
                    }
                    continue;
                }
            };
            for client in &langs {
                let (cargv, cenv) = client.launch(&root);
                let (ok, failures) = run_client(&cargv, &cenv, transport, server.port);
                results.push((
                    transport.clone(),
                    client.name.to_string(),
                    server.name.to_string(),
                    ok,
                    failures,
                ));
            }
            let _ = child.kill();
            let _ = child.wait();
        }
    }

    // Phase 3: report.
    report(&langs, &transports, &results)
}

fn report(
    langs: &[Lang],
    transports: &[String],
    results: &[(String, String, String, bool, Vec<String>)],
) -> Result<()> {
    let cell = |transport: &str, client: &str, server: &str| {
        results
            .iter()
            .find(|(t, c, s, _, _)| t == transport && c == client && s == server)
            .map(|(_, _, _, ok, _)| *ok)
    };

    let mut total = 0usize;
    let mut passed = 0usize;
    for transport in transports {
        println!("\n=== {transport} ===  (rows = client, cols = server)");
        let header: Vec<&str> = langs.iter().map(|l| l.name).collect();
        print!("{:>10} |", "");
        for h in &header {
            print!(" {h:>8}");
        }
        println!();
        for client in langs {
            print!("{:>10} |", client.name);
            for server in langs {
                let mark = match cell(transport, client.name, server.name) {
                    Some(true) => "✓",
                    Some(false) => "✗",
                    None => "·",
                };
                print!(" {mark:>8}");
                total += 1;
                if matches!(cell(transport, client.name, server.name), Some(true)) {
                    passed += 1;
                }
            }
            println!();
        }
    }

    // Failure detail.
    let failures: Vec<_> = results.iter().filter(|(_, _, _, ok, _)| !ok).collect();
    if !failures.is_empty() {
        println!("\n=== failures ===");
        for (t, c, s, _, details) in &failures {
            println!("[{t}] client={c} server={s}");
            for d in details.iter() {
                println!("    - {d}");
            }
        }
    }

    println!("\n== interop summary: {passed}/{total} cells passed ==");
    if passed != total {
        bail!("interop matrix has {} failing cell(s)", total - passed);
    }
    Ok(())
}

File: .reactorcide/scripts/interop.py
"""Build and run the csilgen cross-language interoperability matrix."""

from __future__ import annotations

import argparse
import json
import os
import selectors
import shutil
import subprocess
import sys
import threading
import time
from pathlib import Path
from typing import Dict, List, Mapping, NamedTuple, Sequence, Tuple


class Language(NamedTuple):
    name: str
    generator: str
    package: str
    port: int


LANGUAGES = (
    Language("rust", "rust", "interop_api", 6387),
    Language("go", "go", "interopapi", 6388),
    Language("python", "python", "interop_api", 6389),
    Language("typescript", "typescript", "interop_api", 6390),
    Language("java", "java", "interop_api", 6391),
    Language("csharp", "csharp", "interop_api", 6392),
    Language("c", "c", "interop_api", 6393),
    Language("ruby", "ruby", "interop_api", 6394),
    Language("elixir", "elixir", "interop_api", 6395),
    Language("dart", "dart", "interop_api", 6396),
    Language("ocaml", "ocaml", "interop_api", 6397),
    Language("zig", "zig", "interop_api", 6398),
    Language("kotlin", "kotlin", "interop_api", 6399),
)
TRANSPORTS = ("rpc", "events", "datagrams")


def _run(
    args: Sequence[str | Path],
    cwd: Path,
    *,
    env: Mapping[str, str] | None = None,
    capture: bool = False,
    timeout: float | None = None,
) -> subprocess.CompletedProcess[str]:
    command = tuple(str(arg) for arg in args)
    print("+ " + " ".join(command), flush=True)
    command_env = os.environ.copy()
    if env:
        command_env.update(env)
    return subprocess.run(
        command,
        cwd=cwd,
        env=command_env,
        check=True,
        shell=False,
        text=True,
        capture_output=capture,
        timeout=timeout,
    )


def _harness(root: Path, language: Language) -> Path:
    return root / "tests" / "interop" / "harness" / language.name


def _generated(root: Path, language: Language) -> Path:
    return _harness(root, language) / "gen"


def _build_generators(root: Path, languages: Sequence[Language]) -> None:
    packages = tuple(
        dict.fromkeys(
            f"csilgen-{language.generator}-generator"
            for language in languages
        )
    )
    print("== building interop WASM generators (release) ==", flush=True)
    args = [
        "cargo",
        "build",
        "--release",
        "--quiet",
        "--target",
        "wasm32-unknown-unknown",
    ]
    for package in packages:
        args.extend(("--package", package))
    _run(args, root)


def _generate(root: Path, cli: Path, language: Language) -> None:
    specification = root / "tests" / "interop" / "interop.csil"
    options = (
        "options { emit_packages: "
        f'["{language.generator}"], package_name: "{language.package}", '
        'package_version: "0.1.0" }\n\n'
    )
    input_dir = root / "target" / "interop"
    input_dir.mkdir(parents=True, exist_ok=True)
    generated_input = input_dir / f"gen-input-{language.name}.csil"
    generated_input.write_text(
        options + specification.read_text(encoding="utf-8"),
        encoding="utf-8",
    )
    destination = _generated(root, language)
    if destination.exists():
        shutil.rmtree(destination)
    _run(
        (
            cli,
            "generate",
            "--input",
            generated_input,
            "--target",
            language.generator,
            "--output",
            destination,
        ),
        root,
    )


def _build_ruby(root: Path, harness: Path) -> None:
    for require_path, gem_name in (("bigdecimal", "bigdecimal"), ("json", "json")):
        available = subprocess.run(
            ("ruby", "-e", f'require "{require_path}"'),
            cwd=harness,
            shell=False,
            stdout=subprocess.DEVNULL,
            stderr=subprocess.DEVNULL,
        )
        if available.returncode != 0:
            _run(
                ("gem", "install", "--user-install", "--no-document", gem_name),
                harness,
            )
    _run(
        (
            "ruby",
            f"-I{harness / 'gen' / 'lib'}",
            f"-I{root / 'transports' / 'ruby' / 'lib'}",
            "-e",
            'require "interop_api"; require "csilgen/transport"',
        ),
        harness,
    )


def _build_ocaml(root: Path, harness: Path) -> None:
    transport = harness / "transport"
    if transport.exists():
        shutil.rmtree(transport)
    transport.mkdir(parents=True)
    library = root / "transports" / "ocaml" / "lib"
    for pattern in ("*.ml", "*.mli"):
        for source in sorted(library.glob(pattern)):
            shutil.copy2(source, transport / source.name)
    (transport / "dune").write_text(
        "(library\n (name csilgen_transport)\n (libraries unix))\n",
        encoding="utf-8",
    )
    _run(("dune", "build", "--profile", "release"), harness)


def _build(root: Path, language: Language) -> None:
    harness = _harness(root, language)
    generated = _generated(root, language)
    print(f"  building {language.name} harness...", flush=True)
    if language.name == "rust":
        _run(("cargo", "build", "--quiet"), harness)
    elif language.name == "go":
        _run(
            ("go", "build", "-o", "csil-interop", "."),
            harness,
            env={"GOFLAGS": "-buildvcs=false"},
        )
    elif language.name == "python":
        return
    elif language.name == "typescript":
        _run(("npm", "install", "--no-audit", "--no-fund", "--silent"), generated)
        _run((generated / "node_modules" / ".bin" / "tsc",), generated)
    elif language.name == "java":
        classes = harness / "classes"
        if classes.exists():
            shutil.rmtree(classes)
        classes.mkdir()
        sources = sorted((root / "transports" / "java" / "src" / "main" / "java").rglob("*.java"))
        sources.extend(sorted((generated / "src" / "main" / "java").rglob("*.java")))
        _run(("javac", "-d", classes, *sources, harness / "Main.java"), harness)
    elif language.name == "csharp":
        _run(
            ("dotnet", "build", "csil-interop.csproj", "-c", "Release", "--nologo"),
            harness,
        )
    elif language.name == "c":
        sources = sorted((root / "transports" / "c" / "src").glob("*.c"))
        _run(
            (
                os.environ.get("CC", "cc"),
                "-std=c11",
                "-O2",
                "-Wall",
                "-Wextra",
                f"-I{root / 'transports' / 'c' / 'include'}",
                f"-I{root / 'transports' / 'c' / 'src'}",
                f"-I{harness}",
                harness / "main.c",
                *sources,
                "-o",
                harness / "csil-interop-c",
            ),
            harness,
        )
    elif language.name == "ruby":
        _build_ruby(root, harness)
    elif language.name == "elixir":
        mix_env = {"MIX_ENV": "dev"}
        transport = root / "transports" / "elixir"
        _run(("mix", "deps.get"), transport, env=mix_env)
        _run(("mix", "compile"), transport, env=mix_env)
        _run(("mix", "deps.get"), generated, env=mix_env)
        _run(("mix", "compile"), generated, env=mix_env)
    elif language.name == "dart":
        _run(("dart", "pub", "get"), harness)
    elif language.name == "ocaml":
        _build_ocaml(root, harness)
    elif language.name == "zig":
        _run(("zig", "build"), harness)
    elif language.name == "kotlin":
        _run(("gradle", "--no-daemon", "-q", "installDist"), harness)
    else:
        raise RuntimeError(f"No build command is registered for {language.name}")
    print(f"{language.name} harness built", flush=True)


def _launch(
    root: Path,
    language: Language,
) -> Tuple[Tuple[str, ...], Path, Dict[str, str]]:
    harness = _harness(root, language)
    generated = _generated(root, language)
    if language.name == "rust":
        return ((str(harness / "target" / "debug" / "csil-interop"),), harness, {})
    if language.name == "go":
        return ((str(harness / "csil-interop"),), harness, {})
    if language.name == "python":
        python_path = os.pathsep.join(
            (str(generated), str(root / "transports" / "python"))
        )
        return (("python3", str(harness / "main.py")), harness, {"PYTHONPATH": python_path})
    if language.name == "typescript":
        return (("node", str(harness / "main.ts")), harness, {})
    if language.name == "java":
        return (("java", "-cp", str(harness / "classes"), "interop.Main"), harness, {})
    if language.name == "csharp":
        binary = harness / "bin" / "Release" / "net8.0" / "csil-interop.dll"
        return (("dotnet", str(binary)), harness, {})
    if language.name == "c":
        return ((str(harness / "csil-interop-c"),), harness, {})
    if language.name == "ruby":
        return (
            (
                "ruby",
                f"-I{generated / 'lib'}",
                f"-I{root / 'transports' / 'ruby' / 'lib'}",
                str(harness / "main.rb"),
            ),
            harness,
            {},
        )
    if language.name == "elixir":
        transport_ebin = root / "transports" / "elixir" / "_build" / "dev" / "lib" / "csilgen_transport" / "ebin"
        generated_ebin = generated / "_build" / "dev" / "lib" / "interop_api" / "ebin"
        return (
            (
                "elixir",
                "-pa",
                str(generated_ebin),
                "-pa",
                str(transport_ebin),
                str(harness / "main.exs"),
            ),
            harness,
            {},
        )
    if language.name == "dart":
        return (("dart", "run", "main.dart"), harness, {})
    if language.name == "ocaml":
        return ((str(harness / "_build" / "default" / "main.exe"),), harness, {})
    if language.name == "zig":
        return ((str(harness / "zig-out" / "bin" / "csil-interop-zig"),), harness, {})
    if language.name == "kotlin":
        distribution = harness / "build" / "install" / "csil-interop-kotlin"
        return (
            (
                "java",
                "-cp",
                str(distribution / "lib" / "*"),
                "csil.interop.MainKt",
            ),
            harness,
            {},
        )
    raise RuntimeError(f"No launch command is registered for {language.name}")


def _process_environment(values: Mapping[str, str]) -> Dict[str, str]:
    environment = os.environ.copy()
    environment.update(values)
    return environment


def _drain(stream: object) -> None:
    if stream is None:
        return
    for _ in stream:  # type: ignore[union-attr]
        pass


def _spawn_server(
    root: Path,
    language: Language,
    transport: str,
) -> subprocess.Popen[str]:
    command, cwd, values = _launch(root, language)
    process = subprocess.Popen(
        (*command, "server", transport, str(language.port)),
        cwd=cwd,
        env=_process_environment(values),
        shell=False,
        text=True,
        stdout=subprocess.PIPE,
        stderr=subprocess.DEVNULL,
    )
    if process.stdout is None:
        process.kill()
        raise RuntimeError("The server did not provide a readiness stream")
    selector = selectors.DefaultSelector()
    selector.register(process.stdout, selectors.EVENT_READ)
    deadline = time.monotonic() + 20
    try:
        while True:
            remaining = deadline - time.monotonic()
            if remaining <= 0 or not selector.select(remaining):
                process.kill()
                process.wait()
                raise RuntimeError("server did not become READY within timeout")
            line = process.stdout.readline()
            if not line:
                process.wait()
                raise RuntimeError("server exited before READY")
            if line.strip() == "READY":
                break
    finally:
        selector.close()
    threading.Thread(target=_drain, args=(process.stdout,), daemon=True).start()
    return process


def _run_client(
    root: Path,
    language: Language,
    transport: str,
    port: int,
) -> Tuple[bool, List[str]]:
    command, cwd, values = _launch(root, language)
    try:
        result = subprocess.run(
            (*command, "client", transport, str(port)),
            cwd=cwd,
            env=_process_environment(values),
            shell=False,
            text=True,
            capture_output=True,
            timeout=30,
        )
    except (OSError, subprocess.TimeoutExpired) as error:
        return False, [f"client failed: {error}"]
    json_line = next(
        (line for line in reversed(result.stdout.splitlines()) if line.lstrip().startswith("{")),
        None,
    )
    if json_line is None:
        return False, [f"no JSON result (stderr: {result.stderr.strip()})"]
    try:
        payload = json.loads(json_line)
    except json.JSONDecodeError as error:
        return False, [f"bad JSON: {error}"]
    cases = payload.get("cases")
    if not isinstance(cases, list):
        return False, ["malformed result (no cases array)"]
    if not cases:
        return False, ["no cases run"]
    failures = []
    for case in cases:
        if not isinstance(case, dict) or case.get("ok") is not True:
            name = case.get("name", "?") if isinstance(case, dict) else "?"
            detail = case.get("detail", "") if isinstance(case, dict) else ""
            failures.append(f"{name}: {detail}")
    return not failures, failures


Result = Tuple[str, str, str, bool, List[str]]


def _report(
    languages: Sequence[Language],
    transports: Sequence[str],
    results: Sequence[Result],
) -> None:
    by_cell = {(item[0], item[1], item[2]): item for item in results}
    total = 0
    passed = 0
    for transport in transports:
        print(f"\n=== {transport} ===  (rows = client, cols = server)")
        print(f"{'':>10} |", end="")
        for language in languages:
            print(f" {language.name:>8}", end="")
        print()
        for client in languages:
            print(f"{client.name:>10} |", end="")
            for server in languages:
                item = by_cell.get((transport, client.name, server.name))
                mark = "✓" if item and item[3] else "✗" if item else "·"
                print(f" {mark:>8}", end="")
                total += 1
                if item and item[3]:
                    passed += 1
            print()
    failures = [item for item in results if not item[3]]
    if failures:
        print("\n=== failures ===")
        for transport, client, server, _, details in failures:
            print(f"[{transport}] client={client} server={server}")
            for detail in details:
                print(f"    - {detail}")
    print(f"\n== interop summary: {passed}/{total} cells passed ==")
    if passed != total:
        raise RuntimeError(f"interop matrix has {total - passed} failing cell(s)")


def _select_languages(names: Sequence[str]) -> Tuple[Language, ...]:
    if not names:
        return LANGUAGES
    unknown = sorted(set(names) - {language.name for language in LANGUAGES})
    if unknown:
        raise RuntimeError("Unknown interop languages: " + ", ".join(unknown))
    return tuple(language for language in LANGUAGES if language.name in names)


def _select_transports(names: Sequence[str]) -> Tuple[str, ...]:
    if not names:
        return TRANSPORTS
    unknown = sorted(set(names) - set(TRANSPORTS))
    if unknown:
        raise RuntimeError("Unknown interop transports: " + ", ".join(unknown))
    return tuple(transport for transport in TRANSPORTS if transport in names)


def run(
    root: Path,
    language_names: Sequence[str] = (),
    transport_names: Sequence[str] = (),
) -> None:
    root = root.resolve()
    languages = _select_languages(language_names)
    transports = _select_transports(transport_names)
    cli = root / "target" / "release" / "csilgen"
    _build_generators(root, languages)
    print("== building csilgen CLI (release) ==", flush=True)
    _run(("cargo", "build", "--release", "--quiet", "-p", "csilgen"), root)
    if not cli.is_file():
        raise RuntimeError(f"csilgen binary not found at {cli}")
    print(
        "== interop: langs "
        f"[{', '.join(language.name for language in languages)}] "
        f"transports [{', '.join(transports)}] ==",
        flush=True,
    )
    for language in languages:
        print(f"== preparing {language.name} ==", flush=True)
        _generate(root, cli, language)
        _build(root, language)

    results: List[Result] = []
    for transport in transports:
        for server in languages:
            try:
                process = _spawn_server(root, server, transport)
            except RuntimeError as error:
                for client in languages:
                    results.append(
                        (transport, client.name, server.name, False, [f"server start: {error}"])
                    )
                continue
            try:
                for client in languages:
                    ok, failures = _run_client(
                        root,
                        client,
                        transport,
                        server.port,
                    )
                    results.append((transport, client.name, server.name, ok, failures))
            finally:
                process.kill()
                process.wait()
    _report(languages, transports, results)


def _split_values(value: str) -> Tuple[str, ...]:
    return tuple(item.strip() for item in value.split(",") if item.strip())


def main(argv: Sequence[str] | None = None) -> int:
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("--root", type=Path, default=Path.cwd())
    parser.add_argument("--langs", default="")
    parser.add_argument("--transports", default="")
    args = parser.parse_args(argv)
    try:
        run(
            args.root,
            _split_values(args.langs),
            _split_values(args.transports),
        )
    except (OSError, RuntimeError, subprocess.SubprocessError) as error:
        print(f"interop failed: {error}", file=sys.stderr)
        return 1
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
